core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowSubscript.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.DataType;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.LongLiteral;
+import io.trino.sql.tree.Row;
+import io.trino.sql.tree.RowDataType;
+import io.trino.sql.tree.SubscriptExpression;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * Transforms expressions of the form
+ *
+ * <code>ROW(x, y)[1]</code> to <code>x</code>
+ *
+ * and
+ *
+ * <code>CAST(ROW(x, y) AS row(f1 type1, f2 type2))[1]</code> to <code>CAST(x AS type1)</code>
+ *
+ */
+public class UnwrapRowSubscript
+        extends ExpressionRewriteRuleSet
+{
+    public UnwrapRowSubscript()
+    {
+        super((expression, context) -> ExpressionTreeRewriter.rewriteWith(new Rewriter(), expression));
+    }
+
+    private static class Rewriter
+            extends io.trino.sql.tree.ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression base = treeRewriter.rewrite(node.getBase(), context);
+
+            Deque<Coercion> coercions = new ArrayDeque<>();
+            while (base instanceof Cast) {
+                Cast cast = (Cast) base;
+                if (cast.getType() instanceof RowDataType) {
+                    int index = (int) ((LongLiteral) node.getIndex()).getValue();
+                    base = cast.getExpression();
+                    RowDataType rowType = (RowDataType) cast.getType();
+                    coercions.push(new Coercion(rowType.getFields().get(index - 1).getType(), cast.isTypeOnly(), cast.isSafe()));
+                }
+            }
```

Comment:
won't it loop if the Cast is not into `RowDataType`? E.g. in case of array subscript?

Reply:
Yes, that's why the tests are failing :)